File: lib/records.cc
#include "records.h"
#include "error.h"
#include "primitives.h"
#include "primitives_utils.h"

using namespace net_deserializer;

using NodeFactory = std::function<std::unique_ptr<Node>(BinaryReader &)>;

namespace
{
    enum class RecordType : uint8_t
    {
        SerializedStreamHeader         = 0,
        ClassWithId                    = 1,
        SystemClassWithMembers         = 2,
        ClassWithMembers               = 3,
        SystemClassWithMembersAndTypes = 4,
        ClassWithMembersAndTypes       = 5,
        BinaryObjectString             = 6,
        BinaryArray                    = 7,
        MemberPrimitiveTyped           = 8,
        MemberReference                = 9,
        ObjectNull                     = 10,
        MessageEnd                     = 11,
        BinaryLibary                   = 12,
        ObjectNullMultiple256          = 13,
        ObjectNullMultiple             = 14,
        ArraySinglePrimitive           = 15,
        ArraySingleObject              = 16,
        ArraySingleString              = 17,
        MethodCall                     = 21,
        MethodReturn                   = 22,
    };

    enum class BinaryType : uint8_t
    {
        Primitive      = 0,
        String         = 1,
        Object         = 2,
        SystemClass    = 3,
        Class          = 4,
        ObjectArray    = 5,
        StringArray    = 6,
        PrimitiveArray = 7,
    };

    enum MessageFlags
    {
        NoArgs                 = 0x00000001,
        ArgsInline             = 0x00000002,
        ArgsIsArray            = 0x00000004,
        ArgsInArray            = 0x00000008,
        NoContext              = 0x00000010,
        ContextInline          = 0x00000020,
        ContextInArray         = 0x00000040,
        MethodSignatureInArray = 0x00000080,
        PropertiesInArray      = 0x00000100,
        NoReturnValue          = 0x00000200,
        ReturnValueVoid        = 0x00000400,
        ReturnValueInline      = 0x00000800,
        ReturnValueInArray     = 0x00001000,
        ExceptionInArray       = 0x00002000,
        GenericMethod          = 0x00008000,
    };

    struct MemberReadContext
    {
        std::vector<std::unique_ptr<AggregateNode>> member_nodes;
        std::vector<BinaryType> binary_types;
        std::vector<PrimitiveType> primitive_types;
        std::vector<std::unique_ptr<Node>> additional_info;
    };
}

static MemberReadContext prepare_member_read_context(BinaryReader &reader)
{
    MemberReadContext context;
    const std::size_t member_count = reader.read<int32_t>();
    context.binary_types.resize(member_count);
    context.primitive_types.resize(member_count);
    context.additional_info.resize(member_count);

    for (std::size_t i = 0; i < member_count; i++)
    {
        context.member_nodes.push_back(
            std::make_unique<AggregateNode>("Member"));
    }

    for (std::size_t i = 0; i < member_count; i ++)
    {
        context.member_nodes[i]->add(
            read_primitive("MemberName", reader, PrimitiveType::String));
    }

    for (std::size_t i = 0; i < member_count; i++)
        context.binary_types[i] = reader.read<BinaryType>();

    for (std::size_t i = 0; i < member_count; i++)
    {
        if (context.binary_types[i] == BinaryType::Primitive
            || context.binary_types[i] == BinaryType::PrimitiveArray)
        {
            context.primitive_types[i] = reader.read<PrimitiveType>();
        }
        else if (context.binary_types[i] == BinaryType::SystemClass)
        {
            read_primitive("ClassName", reader, PrimitiveType::String);
        }
        else if (context.binary_types[i] == BinaryType::Class)
        {
            read_primitive("ClassName", reader, PrimitiveType::String);
            read_primitive("LibraryId", reader, PrimitiveType::Int32);
        }
    }
    return context;
}

static std::unique_ptr<Node> read_class_members_with_types(
    BinaryReader &reader, MemberReadContext &context)
{
    for (std::size_t i = 0; i < context.member_nodes.size(); i++)
    {
        if (context.binary_types[i] == BinaryType::Primitive)
        {
            context.member_nodes[i]->add(
                read_primitive("Value", reader, context.primitive_types[i]));
        }
        else if (context.binary_types[i] == BinaryType::String
            || context.binary_types[i] == BinaryType::Object
            || context.binary_types[i] == BinaryType::SystemClass)
        {
            context.member_nodes[i]->add(read_record(reader));
        }
        else
        {
            throw NotImplementedError(
                "Unsupported binary type: " + std::to_string(
                    static_cast<uint8_t>(context.binary_types[i])));
        }
    }

    auto members_node = std::make_unique<AggregateNode>("Members");
    for (auto &member_node : context.member_nodes)
        members_node->add(std::move(member_node));
    return members_node;
}

static std::unique_ptr<Node> read_serialized_stream_header(BinaryReader &reader)
{
    auto node = std::make_unique<AggregateNode>("SerializedStreamHeader");
    node->add(read_primitive("RootId", reader, PrimitiveType::Int32));
    node->add(read_primitive("HeaderId", reader, PrimitiveType::Int32));
    node->add(read_primitive("MajorVersion", reader, PrimitiveType::Int32));
    node->add(read_primitive("MinorVersion", reader, PrimitiveType::Int32));
    return node;
}

static std::unique_ptr<Node> read_system_class_with_members(
    BinaryReader &reader)
{
    throw NotImplementedError("System class with members");
}

static std::unique_ptr<Node> read_class_with_members(
    BinaryReader &reader)
{
    throw NotImplementedError("Class with members");
}

static std::unique_ptr<Node> read_system_class_with_members_and_types(
    BinaryReader &reader)
{
    auto node = std::make_unique<AggregateNode>("SystemObject");
    node->add(read_primitive("ObjectId", reader, PrimitiveType::Int32));
    node->add(read_primitive("ObjectName", reader, PrimitiveType::String));
    auto member_read_context = prepare_member_read_context(reader);
    // node->add(read_primitive("LibraryId", reader, PrimitiveType::Int32));
    node->add(read_class_members_with_types(reader, member_read_context));
    return node;
}

static std::unique_ptr<Node> read_class_with_members_and_types(
    BinaryReader &reader)
{
    auto node = std::make_unique<AggregateNode>("Object");
    node->add(read_primitive("ObjectId", reader, PrimitiveType::Int32));
    node->add(read_primitive("ObjectName", reader, PrimitiveType::String));
    auto member_read_context = prepare_member_read_context(reader);
    node->add(read_primitive("LibraryId", reader, PrimitiveType::Int32));
    node->add(read_class_members_with_types(reader, member_read_context));
    return node;
}

static std::unique_ptr<Node> read_binary_object_string(BinaryReader &reader)
{
    auto node = std::make_unique<AggregateNode>("StringObject");
    node->add(read_primitive("ObjectId", reader, PrimitiveType::Int32));
    node->add(read_primitive("Value", reader, PrimitiveType::String));
    return node;
}

static std::unique_ptr<Node> read_member_reference(BinaryReader &reader)
{
    return read_primitive("ObjectReference", reader, PrimitiveType::Int32);
}

static std::unique_ptr<Node> read_object_null(BinaryReader &reader)
{
    return std::make_unique<LeafNode>("NullObject");
}

static std::unique_ptr<Node> read_message_end(BinaryReader &reader)
{
    return std::make_unique<LeafNode>("MessageEnd");
}

static std::unique_ptr<Node> read_binary_library(BinaryReader &reader)
{
    auto node = std::make_unique<AggregateNode>("BinaryLibrary");
    node->add(read_primitive("LibraryId", reader, PrimitiveType::Int32));
    node->add(read_primitive("LibraryName", reader, PrimitiveType::String));
    return node;
}

static std::unique_ptr<Node> read_array_single_primitive(BinaryReader &reader)
{
    throw NotImplementedError("Array of single primitive");
}

static std::unique_ptr<Node> read_array_single_object(BinaryReader &reader)
{
    auto node = std::make_unique<AggregateNode>("Array");
    auto elements_node = std::make_unique<AggregateNode>("Elements");
    node->add(read_primitive("ObjectId", reader, PrimitiveType::Int32));
    const auto length = reader.read<uint32_t>();
    for (std::size_t i = 0; i < length; i++)
        elements_node->add(read_record(reader));
    node->add(std::move(elements_node));
    return node;
}

static std::unique_ptr<Node> read_array_of_value_with_code(BinaryReader &reader)
{
    throw NotImplementedError("Array of value with code");
}

static std::unique_ptr<Node> read_array_single_string(BinaryReader &reader)
{
    throw NotImplementedError("Array of single string");
}

static std::unique_ptr<Node> read_binary_method_call(BinaryReader &reader)
{
    auto node = std::make_unique<AggregateNode>("BinaryMethodCall");
    const auto flags = reader.read<uint32_t>();
    node->add(std::make_unique<LeafNode>("Flags", flags));
    node->add(read_primitive("MethodName", reader));
    node->add(read_primitive("TypeName", reader));
    if (flags & MessageFlags::ContextInline)
        node->add(read_primitive("CallContext", reader));
    if (flags & MessageFlags::ArgsInline)
    {
        auto args_node = read_array_of_value_with_code(reader);
        args_node->name = "Args";
        node->add(std::move(args_node));
    }
    else if (!(flags & MessageFlags::NoArgs))
    {
        auto args_node = read_record(reader);
        args_node->name = "Args";
        node->add(std::move(args_node));
    }
    return std::move(node);
}

static NodeFactory read_stub(const RecordType record_type)
{
    return [=](BinaryReader &) -> std::unique_ptr<Node>
    {
        throw NotImplementedError(
            "Reading record type "
            + std::to_string(static_cast<int>(record_type))
            + " is not yet implemented");
    };
}

using RT = RecordType;
static std::map<RecordType, NodeFactory> node_factory_map =
{
    {RT::SerializedStreamHeader,         read_serialized_stream_header},
    {RT::ClassWithId,                    read_stub(RT::ClassWithId)},
    {RT::SystemClassWithMembers,         read_system_class_with_members},
    {RT::ClassWithMembers,               read_class_with_members},
    {RT::SystemClassWithMembersAndTypes, read_system_class_with_members_and_types},
    {RT::ClassWithMembersAndTypes,       read_class_with_members_and_types},
    {RT::BinaryObjectString,             read_binary_object_string},
    {RT::BinaryArray,                    read_stub(RT::BinaryArray)},
    {RT::MemberPrimitiveTyped,           read_stub(RT::MemberPrimitiveTyped)},
    {RT::MemberReference,                read_member_reference},
    {RT::ObjectNull,                     read_object_null},
    {RT::MessageEnd,                     read_message_end},
    {RT::BinaryLibary,                   read_binary_library},
    {RT::ObjectNullMultiple256,          read_stub(RT::ObjectNullMultiple256)},
    {RT::ObjectNullMultiple,             read_stub(RT::ObjectNullMultiple)},
    {RT::ArraySinglePrimitive,           read_array_single_primitive},
    {RT::ArraySingleObject,              read_array_single_object},
    {RT::ArraySingleString,              read_array_single_string},
    {RT::MethodCall,                     read_binary_method_call},
    {RT::MethodReturn,                   read_stub(RT::MethodReturn)},
};

std::unique_ptr<Node> net_deserializer::read_record(BinaryReader &reader)
{
    const RecordType record_type = reader.read<RecordType>();
    if (node_factory_map.find(record_type) == node_factory_map.end())
    {
        throw NotImplementedError("Unknown record type: "
            + std::to_string(static_cast<int>(record_type)));
    }
    return node_factory_map[record_type](reader);
}

File: net_deserializer/records.py
from enum import Enum
from net_deserializer import util
from net_deserializer.node import Node, LeafNode
from net_deserializer.primitives import read_primitive, PrimitiveType

class CorruptDataError(RuntimeError):
    pass

class RecordType(Enum):
    # pylint: disable=bad-whitespace
    SerializedStreamHeader         = 0
    ClassWithId                    = 1
    SystemClassWithMembers         = 2
    ClassWithMembers               = 3
    SystemClassWithMembersAndTypes = 4
    ClassWithMembersAndTypes       = 5
    BinaryObjectString             = 6
    BinaryArray                    = 7
    MemberPrimitiveTyped           = 8
    MemberReference                = 9
    ObjectNull                     = 10
    MessageEnd                     = 11
    BinaryLibrary                  = 12
    ObjectNullMultiple256          = 13
    ObjectNullMultiple             = 14
    ArraySinglePrimitive           = 15
    ArraySingleObject              = 16
    ArraySingleString              = 17
    MethodCall                     = 21
    MethodReturn                   = 22

class BinaryType(Enum):
    # pylint: disable=bad-whitespace
    Primitive      = 0
    String         = 1
    Object         = 2
    SystemClass    = 3
    Class          = 4
    ObjectArray    = 5
    StringArray    = 6
    PrimitiveArray = 7

class BinaryArrayType(Enum):
    # pylint: disable=bad-whitespace
    Single            = 0
    Jagged            = 1
    Rectangular       = 2
    SingleOffset      = 3
    JaggedOffset      = 4
    RectangularOffset = 5

class MessageFlags(Enum):
    # pylint: disable=bad-whitespace
    NoArgs                 = 0x00000001
    ArgsInline             = 0x00000002
    ArgsIsArray            = 0x00000004
    ArgsInArray            = 0x00000008
    NoContext              = 0x00000010
    ContextInline          = 0x00000020
    ContextInArray         = 0x00000040
    MethodSignatureInArray = 0x00000080
    PropertiesInArray      = 0x00000100
    NoReturnValue          = 0x00000200
    ReturnValueVoid        = 0x00000400
    ReturnValueInline      = 0x00000800
    ReturnValueInArray     = 0x00001000
    ExceptionInArray       = 0x00002000
    GenericMethod          = 0x00008000

class Context():
    def __init__(self, handle):
        self.handle = handle
        self.objects = {}

def _is_class(record_type):
    return record_type in (
        RecordType.ClassWithId,
        RecordType.ClassWithMembers,
        RecordType.ClassWithMembersAndTypes,
        RecordType.SystemClassWithMembers,
        RecordType.SystemClassWithMembersAndTypes)

def _is_array(record_type):
    return record_type in(
        RecordType.BinaryArray,
        RecordType.ArraySinglePrimitive,
        RecordType.ArraySingleObject,
        RecordType.ArraySingleString)

def _read_create_object(ctx, name):
    object_id = read_primitive(ctx.handle, PrimitiveType.Int32)
    node = Node(name)
    ctx.objects[object_id] = node
    node.add_leaf('ObjectId', object_id)
    return node, object_id

def _read_binary_type(ctx):
    return BinaryType(util.read_fmt(ctx.handle, 'B'))

def _read_record_type(ctx):
    return RecordType(util.read_fmt(ctx.handle, 'B'))

def _read_primitive_type(ctx):
    return PrimitiveType(util.read_fmt(ctx.handle, 'B'))

def _read_from_binary_type(ctx, binary_type, primitive_type):
    if binary_type == BinaryType.Primitive:
        return LeafNode('Value', read_primitive(ctx.handle, primitive_type))

    if binary_type in (
            BinaryType.String,
            BinaryType.Object,
            BinaryType.PrimitiveArray,
            BinaryType.StringArray,
            BinaryType.ObjectArray,
            BinaryType.Class,
            BinaryType.SystemClass):
        record_type = _read_record_type(ctx)
        return _read_record(ctx, record_type)

    raise NotImplementedError(
        'Unimplemented binary type: {0}'.format(binary_type))

def _read_class_members_meta(ctx, object_id):
    count = read_primitive(ctx.handle, PrimitiveType.Int32)
    names = [None] * count
    binary_types = [None] * count
    for i in range(count):
        names[i] = read_primitive(ctx.handle, PrimitiveType.String)
    for i in range(count):
        binary_types[i] = _read_binary_type(ctx)

    node = Node('MembersMeta')
    for i in range(count):
        member_node = Node('MemberMeta')
        member_node.add_leaf('Name', names[i])
        member_node.add_leaf('BinaryType', binary_types[i])
        if binary_types[i] in (BinaryType.Primitive, BinaryType.PrimitiveArray):
            member_node.add_leaf('PrimitiveType', _read_primitive_type(ctx))
        elif binary_types[i] == BinaryType.SystemClass:
            member_node.add_leaf(
                'ClassName', read_primitive(ctx.handle, PrimitiveType.String))
        elif binary_types[i] == BinaryType.Class:
            member_node.add_leaf(
                'ClassName', read_primitive(ctx.handle, PrimitiveType.String))
            member_node.add_leaf(
                'LibraryId', read_primitive(ctx.handle, PrimitiveType.Int32))
        node.add(member_node)
    return node

def _read_class_members(ctx, object_id):
    if not object_id in ctx.objects:
        raise CorruptDataError(
            'Bad reference to ObjectID {0}'.format(object_id))
    members_meta = ctx.objects[object_id].get('MembersMeta')
    member_nodes = Node('Members')
    for i, member_meta in enumerate(members_meta.children):
        member_node = Node('Member')
        member_node.add(member_meta.get('Name'))
        member_node.add(_read_from_binary_type(
            ctx,
            member_meta.get_leaf('BinaryType'),
            member_meta.get_leaf('PrimitiveType', None)))
        member_nodes.add(member_node)
    return member_nodes

def _read_serialized_stream_header(ctx):
    node = Node('SerializedStreamHeader')
    node.add_leaf('RootId', read_primitive(ctx.handle, PrimitiveType.Int32))
    node.add_leaf('HeaderId', read_primitive(ctx.handle, PrimitiveType.Int32))
    node.add_leaf('MajorVersion', read_primitive(ctx.handle, PrimitiveType.Int32))
    node.add_leaf('MinorVersion', read_primitive(ctx.handle, PrimitiveType.Int32))
    return node

def _read_class_with_id(ctx):
    node, object_id = _read_create_object(ctx, 'ClassWithId')
    metadata_id = util.read_fmt(ctx.handle, '<l')
    node.add(ctx.objects[metadata_id].get('MembersMeta'))
    node.add(_read_class_members(ctx, metadata_id))
    return node

def _read_system_class_with_members_and_types(ctx):
    node, object_id = _read_create_object(ctx, 'SystemClassWithMembersAndTypes')
    node.add_leaf('ObjectName', read_primitive(ctx.handle, PrimitiveType.String))
    node.add(_read_class_members_meta(ctx, object_id))
    node.add(_read_class_members(ctx, object_id))
    return node

def _read_class_with_members_and_types(ctx):
    node, object_id = _read_create_object(ctx, 'ClassWithMembersAndTypes')
    node.add_leaf('ObjectName', read_primitive(ctx.handle, PrimitiveType.String))
    node.add(_read_class_members_meta(ctx, object_id))
    node.add_leaf('LibraryId', read_primitive(ctx.handle, PrimitiveType.Int32))
    node.add(_read_class_members(ctx, object_id))
    return node

def _read_binary_object_string(ctx):
    node, object_id = _read_create_object(ctx, 'BinaryObjectString')
    node.add_leaf('Value', read_primitive(ctx.handle, PrimitiveType.String))
    return node

def _read_binary_array(ctx):
    node, object_id = _read_create_object(ctx, 'BinaryArray')
    array_type = BinaryArrayType(util.read_fmt(ctx.handle, 'B'))
    node.add_leaf('BinaryArrayType', array_type)

    rank = util.read_fmt(ctx.handle, '<l')
    node.add_leaf('Rank', rank)

    dimensions_node = Node('Dimensions')
    element_count = 0
    for i in range(rank):
        dimension = util.read_fmt(ctx.handle, '<l')
        dimensions_node.add_leaf('Dimension', dimension)
        if i == 0:
            element_count = dimension
        else:
            element_count *= dimension
    node.add(dimensions_node)

    if array_type in (
            BinaryArrayType.SingleOffset,
            BinaryArrayType.JaggedOffset,
            BinaryArrayType.RectangularOffset):
        lower_bounds_node = Node('LowerBounds')
        for i in range(rank):
            lower_bounds_node.add_leaf(
                'LowerBound', util.read_fmt(ctx.handle, '<l'))
        node.add(lower_bounds_node)

    binary_type = _read_binary_type(ctx)
    primitive_type = PrimitiveType.Invalid
    if binary_type in (BinaryType.Primitive, BinaryType.PrimitiveArray):
        primitive_type = _read_primitive_type(ctx)
    elif binary_type == BinaryType.SystemClass:
        node.add_leaf('ClassName', read_primitive(ctx.handle, PrimitiveType.String))
    elif binary_type == BinaryType.Class:
        node.add_leaf('ClassName', read_primitive(ctx.handle, PrimitiveType.String))
        node.add_leaf('LibraryId', read_primitive(ctx.handle, PrimitiveType.Int32))

    elements_node = Node('Elements')
    for i in range(element_count):
        elements_node.add(_read_from_binary_type(ctx, binary_type, primitive_type))
    node.add(elements_node)
    return node

def _read_member_primitive_typed(ctx):
    primitive_type = _read_primitive_type(ctx)
    return LeafNode(
        'MemberPrimitiveTyped',
        read_primitive(ctx.handle, primitive_type))

def _read_member_reference(ctx):
    return LeafNode(
        'MemberReference',
        read_primitive(ctx.handle, PrimitiveType.Int32))

def _read_object_null(ctx):
    return LeafNode('NullObject')

def _read_message_end(ctx):
    return LeafNode('MessageEnd')

def _read_binary_library(ctx):
    library_node = Node('BinaryLibrary')
    library_node.add_leaf('LibraryId', read_primitive(ctx.handle, PrimitiveType.Int32))
    library_node.add_leaf('LibraryName', read_primitive(ctx.handle, PrimitiveType.String))
    record_type = _read_record_type(ctx)
    if _is_class(record_type):
        library_node.add(_read_record(ctx, record_type))
        return library_node
    elif _is_array(record_type):
        library_node.add(_read_record(ctx, record_type))
        return library_node
    else:
        raise CorruptDataError(
            'Binary library followed with neither class or array')

def _read_array_single_primitive(ctx):
    node, object_id = _read_create_object(ctx, 'ArraySinglePrimitive')
    length = util.read_fmt(ctx.handle, '<l')
    primitive_type = _read_primitive_type(ctx)
    elements_node = Node('Elements')
    for i in range(length):
        elements_node.add_leaf('Element', read_primitive(ctx.handle, primitive_type))
    node.add(elements_node)
    return node

def _read_array_single_string(ctx):
    node, object_id = _read_create_object(ctx, 'ArraySingleString')
    length = util.read_fmt(ctx.handle, '<L')
    elements_node = Node('Elements')
    for i in range(length):
        record_type = _read_record_type(ctx)
        elements_node.add_leaf('Element', _read_record(ctx, record_type))
    node.add(elements_node)
    return node

def _read_array_single_object(ctx):
    node, object_id = _read_create_object(ctx, 'ArraySingleObject')
    length = util.read_fmt(ctx.handle, '<l')
    record_type = _read_record_type(ctx)
    elements_node = Node('Elements')
    for i in range(length):
        elements_node.add_leaf('Element', _read_record(ctx, record_type))
    node.add(elements_node)
    return node

def _read_array_of_value_with_code(ctx):
    raise NotImplementedError('Unimplemented: array of value with code')

def _read_method_call(ctx):
    node = Node('BinaryMethodCall')

    flags = util.read_fmt(ctx.handle, '<L')
    node.add_leaf('Flags', flags)

    primitive_type = _read_primitive_type(ctx)
    node.add_leaf('MethodName', read_primitive(ctx.handle, primitive_type))

    primitive_type = _read_primitive_type(ctx)
    node.add_leaf('TypeName', read_primitive(ctx.handle, primitive_type))

    if (flags & MessageFlags.ContextInline.value) != 0:
        primitive_type = _read_primitive_type(ctx)
        node.add_leaf('CallContext', read_primitive(ctx.handle, primitive_type))

    if (flags & MessageFlags.ArgsInline.value) != 0:
        args_node = _read_array_of_value_with_code(ctx)
        args_node.name = 'Args'
        node.add(args_node)
    elif (flags & MessageFlags.NoArgs.value) == 0:
        record_type = _read_record_type(ctx)
        args_node = _read_record(ctx, record_type)
        args_node.name = 'Args'
        node.add(args_node)

    return node

def _read_record(ctx, record_type):
    try:
        return _NODE_FACTORY_MAP[record_type](ctx)
    except KeyError:
        raise NotImplementedError(
            'Unimplemented record type: {0}'.format(record_type))

_NODE_FACTORY_MAP = {
    RecordType.SerializedStreamHeader:         _read_serialized_stream_header,
    RecordType.ClassWithId:                    _read_class_with_id,
    RecordType.SystemClassWithMembersAndTypes: _read_system_class_with_members_and_types,
    RecordType.ClassWithMembersAndTypes:       _read_class_with_members_and_types,
    RecordType.BinaryObjectString:             _read_binary_object_string,
    RecordType.BinaryArray:                    _read_binary_array,
    RecordType.MemberPrimitiveTyped:           _read_member_primitive_typed,
    RecordType.MemberReference:                _read_member_reference,
    RecordType.ObjectNull:                     _read_object_null,
    RecordType.MessageEnd:                     _read_message_end,
    RecordType.BinaryLibrary:                  _read_binary_library,
    RecordType.ArraySinglePrimitive:           _read_array_single_primitive,
    RecordType.ArraySingleString:              _read_array_single_string,
    RecordType.ArraySingleObject:              _read_array_single_object,
    RecordType.MethodCall:                     _read_method_call,
}

def deserialize(handle):
    ctx = Context(handle)
    root = Node('Root')
    while True:
        try:
            record_type = _read_record_type(ctx)
        except EOFError:
            break
        child = _read_record(ctx, record_type)
        root.add(child)
    return root
